### ISSUE TITLE
AsyncDns hang/deadlock causing permanent timeouts resolving affected names.

### ISSUE THREAD

**Issue by nodefactory-bk** [akka/akka#32109]
We have run into an issue in our production env where async dns seems to deadlock on some domain names.

When this happens those names will never resolve again and all requests to do so will time out. 
Only way to fix is to restart.

We do not use akka-dns directly, we use akka-http and when this happens all requests to such a name will time out every time until the application is restarted.

Logs show attempt to resolve:
```
Resolution request for bad1.example.com Ip(true,true) from Actor[akka://sys-prod/system/IO-TCP/selectors/$a/70814#-1379895057]
```
But then nothing else is logged regarding that name from the dns actors.

Which names are affected is completely random.

This problem appeared after switching from akka 2.6.20 to 2.8.4. Although I cannot definitely say it never happened before I've looked through some old logs and can't find an earlier instance.Attached are two log files, one showing a domain failing. The issue starts 2023-09-12 16:06:45.828. (bad.log)
The other is from a working name during the same timeframe. (good.log)

[bad1.log](https://github.com/akka/akka/files/12657209/bad1.log)
[good1.log](https://github.com/akka/akka/files/12657210/good1.log)

I note that there were two changes to async dns in akka 2.8.1: #31906 and #31926 .
I do see in #31926 that there changes made to the caching code and since caches really attract these kinds of issues I'd say that's a good place to start.

I would like to stress that we are unable to reproduce this in a test env. It happens only in a very busy production env and rarely at that, however it does cause us operational problems since it can't be solved without restarting so I would suggest this is a pretty critical issue.

**Comment by leviramsey** [comment]
Is it possible to include logs from the `DnsResolutionActor` and `RequestIdInjector`?

I note that in the bad log that just over 5 seconds elapses from `Message [9875] to [/10.15.240.10:53]:` to `Decoded UDP DNS response [Message(9875...`, which would exceed the default `akka.io.dns.async-dns.resolve-timeout`, so the issue would appear to be in the timeout handling.

**Comment by nodefactory-bk** [comment]
Attached is the same log as bad1 above but also containing `DnsResolutionActor`
I couldn't find anything at all matching `RequestIdInjector`.

As you say it is a timeout:
```
2023-09-12 16:06:45.698 Resolve of bad1.example.com timed out after 5.000 s.  Trying next name server akka.io.dns.internal.AsyncDnsResolver$DnsResolutionActor
```

[bad1-b.log](https://github.com/akka/akka/files/12669480/bad1-b.log)

Will this fix be included in akka 2.8.5 and when can we expect that version to be published?
Alternatively are there snapshot versions available somewhere we could switch to in the mean time?

**Comment by nodefactory-bk** [comment]
I have looked through our logs and can confirm that all cases where we've seen this issue it started after a timeout.

Looking through all timeouts in the logs about 50% of them seems to have caused an issue with stuck domain names, however I only found 7 such timeouts so not really enough of a sample set to determine how often it happens.
